### DISCUSSION TITLE
Implementing inversion for all spin components

### DISCUSSION THREAD

**Discussion by miflynn** [QuSpin/QuSpin#639] [Q&A] [answered]
The QuSpin documentation regularly makes use of inversion and parity symmetries to showcase the power of the package. However, I'm curious to learn how best to implement other simple symmetries which are closely related to these examples - in particular, inversion of x- and y-components of spin.

An example of where this can appear is the XYZ model of two spin-s degrees of freedom. Any model of that type is invariant under parity and inversion about any Cartesian axis, and all of the symmetries commute. It's easy to work out how x- and y-inversion act on the integer representation of states in this model, but the process of programming in such a symmetry seems complicated (from my reading of the documentation, the only way to do this is through the user_basis class).

Since each inversion operation is only defined on-site, is there a simple way to introduce these symmetries, or do I need to figure out how to work with user_basis with an arbitrary local Hilbert space dimension?

**Comment by mgbukov** [comment, marked as answer]
In general with `quspin` you should be able to implement _any_ symmetry,
provided its action on the states in the integer representation has
_real-valued_ coefficients.

For example, the inversion of the x- and z-component are such symmetries,
but the inversion of the y-component alone is not. However, note that in
some cases (i) it might be possible to rotate your basis so that the
y-inversion now becomes an x- or z-inversion; (ii) if the symmetry
operation contains an even number of y-flips, the coefficients of the
resulting states are real-valued so this can be implemented.

The proper tool to implement these symmetries is the `user_basis`, see this
[tutorial](https://quspin.github.io/QuSpin/user_basis.html). As a
nontrivial example of how to implement a custom symmetry using the
`user_basis`, take a look at implementing the [Z2 symmetry of plaquette
operators XYZXYZ on the honeycomb lattice](
https://quspin.github.io/QuSpin/examples/example28.html#example28-label).

Finally, it is possible to implement imaginary-valued symmetry operations,
and we have ideas on how to do this, so if you're versed in using cpp and
want to go for it, we'd be happy to support your efforts (and then
incorporate your contribution to a future `quspin` release).

On Sat, Aug 5, 2023 at 10:02 PM miflynn ***@***.***> wrote:

> The QuSpin documentation regularly makes use of inversion and parity
> symmetries to showcase the power of the package. However, I'm curious to
> learn how best to implement other simple symmetries which are closely
> related to these examples - in particular, inversion of x- and y-components
> of spin.
>
> An example of where this can appear is the XYZ model of two spin-s degrees
> of freedom. Any model of that type is invariant under parity and inversion
> about any Cartesian axis, and all of the symmetries commute. It's easy to
> work out how x- and y-inversion act on the integer representation of states
> in this model, but the process of programming in such a symmetry seems
> complicated (from my reading of the documentation, the only way to do this
> is through the user_basis class).
>
> Since each inversion operation is only defined on-site, is there a simple
> way to introduce these symmetries, or do I need to figure out how to work
> with user_basis with an arbitrary local Hilbert space dimension?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/QuSpin/QuSpin/discussions/639>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHSUBW3JTLKZZ7FN75XHUDXT2RGTANCNFSM6AAAAAA3FNAABI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Marin Bukov, PhD
Research Group Leader
Max Planck Institute for the Physics of Complex Systems
Noethnitzer Str. 38
Dresden 01187, Germany

url: https://www.pks.mpg.de/nqd
<http://quantum-dynamics.phys.uni-sofia.bg/>

**Reply by miflynn** [reply]
Thanks for the answer, this was very helpful - I didn't realize that the `sign_ptr` variable could be used in the way described in the Kitaev honeycomb example. Indeed for the problem I'm currently working on I can find a way to implement y-inversion in a pairwise fashion so this is already sufficient - much appreciated!

**Comment by mgbukov** [comment]
We also didn't think of using the sign pointer in that way when it was
first introduced; that's why quspin currently can't handle a generic
complex valued symmetry. The generalization of the code I was talking about
then includes making the sign pointer accept complex phases. Then quspin
would also be able to do anyonic statistics.

Btw, once you have your preprint on the arXiv, we'd be happy to show this
2-spin Hamiltonian example together with the other quspin examples on the
documentation webpage. If you're interested in contributing, let me know
then and I'll give you more info on how to format it, etc.
